src/engine/tests/Test_MILPEncoder.h

```diff
@@ -890,14 +889,188 @@ class MILPEncoderTestSuite : public CxxTest::TestSuite
+        inputQuery.setLowerBound( 2, -10 );
+        inputQuery.setUpperBound( 2, 0.5 );
+        tableau.setLowerBound( 2, -10 );
+        tableau.setUpperBound( 2, 0.5 );
+        inputQuery.setLowerBound( 3, -10 );
+        inputQuery.setUpperBound( 3, 10 );
+        tableau.setLowerBound( 3, -10 );
+        tableau.setUpperBound( 3, 10 );
+
+        BilinearConstraint *bilinear1 = new BilinearConstraint( 0, 1, 2 );
+        inputQuery.addNonlinearConstraint( bilinear1 );
+        BilinearConstraint *bilinear2 = new BilinearConstraint( 2, 1, 3 );
+        inputQuery.addNonlinearConstraint( bilinear2 );
+
+        MILPEncoder milp( tableau );
+        milp.encodeInputQuery( gurobi, inputQuery );
+
+        TS_ASSERT_THROWS_NOTHING( gurobi.solve() );
+
+        TS_ASSERT( !gurobi.haveFeasibleSolution() );
+        TS_ASSERT( gurobi.infeasible() );
+#else
+        TS_ASSERT( true );
+#endif // ENABLE_GUROBI
+    }
+
+    void test_encode_softmax_constraint()
+    {
+#ifdef ENABLE_GUROBI
+
+        /*
+          1.5 <= x0 <= 2
+          0 <= x1 <= 0.5
+        */
+        GurobiWrapper gurobi;
+
+        InputQuery inputQuery = InputQuery();
+        inputQuery.setNumberOfVariables( 4 );
+
+        SoftmaxConstraint *softmax = new SoftmaxConstraint( {0, 1}, {2, 3} );
+        softmax->notifyLowerBound( 0, 1.5 );
+        softmax->notifyUpperBound( 0, 2 );
+        softmax->notifyLowerBound( 1, 0 );
+        softmax->notifyUpperBound( 1, 0.5 );
+
+        MockTableau tableau = MockTableau();
+        tableau.setDimensions( 2, 2 );
+        inputQuery.setLowerBound( 0, 1.5 );
+        tableau.setLowerBound( 0, 1.5 );
+        inputQuery.setUpperBound( 0, 2 );
+        tableau.setUpperBound( 0, 2 );
+        inputQuery.setLowerBound( 1, 0 );
+        tableau.setLowerBound( 1, 0.5 );
+        inputQuery.setUpperBound( 1, 0 );
+        tableau.setUpperBound( 1, 0.5 );
+        softmax->registerTableau( &tableau );
+
+        List<Tightening> tightenings;
+        softmax->getEntailedTightenings( tightenings );
+
+        for ( const auto &t : tightenings )
+        {
+            if ( t._type == Tightening::LB )
+            {
+                inputQuery.setLowerBound( t._variable, t._value );
+                tableau.setLowerBound( t._variable, t._value );
+            }
+            if ( t._type == Tightening::UB )
+            {
+                inputQuery.setUpperBound( t._variable, t._value );
+                tableau.setUpperBound( t._variable, t._value );
+            }
+        }
+        inputQuery.addNonlinearConstraint( softmax );
+
+        MILPEncoder milp( tableau );
+        TS_ASSERT_THROWS_NOTHING( milp.encodeInputQuery( gurobi, inputQuery ) );
+
+        TS_ASSERT_THROWS_NOTHING( gurobi.solve() );
+
+        TS_ASSERT( gurobi.haveFeasibleSolution() );
+
+        Map<String, double> solution;
+        double costValue;
+        TS_ASSERT_THROWS_NOTHING( gurobi.extractSolution( solution, costValue ) );
```

Comment:
Isn't possible to check if values of solution are expected ones?

Reply:
I think in this case we do not have guarantees about the actual assignment. Since we encode the relaxed softmax.